UniversalAlgebra/Structures/Homs.lagda

```diff
@@ -0,0 +1,528 @@
+                                                           ; module ≡-Reasoning )
+
+
+
+
+
+open import Overture.Preliminaries   using ( ∣_∣ ; ∥_∥ ; _∙_ ; _⁻¹)
+open import Overture.Inverses        using ( IsInjective ; IsSurjective )
+open import Relations.Discrete       using ( _|:_ ; 0[_] ; ker )
+open import Relations.Quotients      using ( Equivalence ; Quotient
+                                           ; 0[_]Equivalence ; ker-IsEquivalence ; kerlift-IsEquivalence
+                                           ; ⟪_⟫ ; ⌞_⌟ ; ⟪_∼_⟫-elim ; _/_ )
+open import Relations.Extensionality using ( swelldef )
+
+open import Structures.Basic         using ( Signature ; Structure ; _ʳ_ ; _ᵒ_ ; Compatible
+                                           ; Lift-Strucʳ ; Lift-Strucˡ ; Lift-Struc )
+open import Structures.Congruences
+
+
+private variable 𝑅 𝐹 : Signature
+
+
+-- Development for Structures (Sigma type representation)
+
+module _ {α ρ β ρ' : Level}
+         (𝑨 : Structure {α} {ρ} 𝑅 𝐹)
+         (𝑩 : Structure {β} {ρ'} 𝑅 𝐹) where
+
+ CompRel : (fst 𝑅) → ((fst 𝑨) → (fst 𝑩)) → Type (α ⊔ ρ ⊔ ρ')
+ CompRel R h = ∀ a → ((R ʳ 𝑨) a) → ((R ʳ 𝑩) (h ∘ a))
+
+ IsHom-rel : ((fst 𝑨) → (fst 𝑩)) → Type (α ⊔ ρ ⊔ ρ')
+ IsHom-rel h = ∀ R →  CompRel R h
+
+ CompOp : (fst 𝐹) → ((fst 𝑨) → (fst 𝑩)) → Type (α ⊔ β)
+ CompOp f h = ∀ a → h ((f ᵒ 𝑨) a) ≡ (f ᵒ 𝑩) (h ∘ a)
+
+ IsHom-op : ((fst 𝑨) → (fst 𝑩)) → Type (α ⊔ β)
+ IsHom-op h = ∀ f → CompOp f h
+
+ IsHom : ((fst 𝑨) → (fst 𝑩)) → Type (α ⊔ ρ ⊔ β ⊔ ρ')
+ IsHom h = IsHom-rel h × IsHom-op h
+
+ Hom : Type (α ⊔ ρ ⊔ β ⊔ ρ')
+ Hom = Σ[ h ∈ ((fst 𝑨) → (fst 𝑩)) ] IsHom h
+
+
+
+module _ {α ρᵃ β ρᵇ γ ρᶜ : Level}{𝑅 𝐹 : Signature}
+         (𝑨 : Structure {α} {ρᵃ} 𝑅 𝐹){𝑩 : Structure {β}{ρᵇ} 𝑅 𝐹}
+         (𝑪 : Structure {γ} {ρᶜ} 𝑅 𝐹) where
+
+ ∘-IsHom-rel : {f : (fst 𝑨) → (fst 𝑩)}{g : (fst 𝑩) → (fst 𝑪)}
+  →             IsHom-rel 𝑨 𝑩 f → IsHom-rel 𝑩 𝑪 g → IsHom-rel 𝑨 𝑪 (g ∘ f)
+ ∘-IsHom-rel {f}{g} fhr ghr R a = λ z → ghr R (λ z₁ → f (a z₁)) (fhr R a z)
+
+ ∘-IsHom-op : {f : (fst 𝑨) → (fst 𝑩)}{g : (fst 𝑩) → (fst 𝑪)}
+  →            IsHom-op 𝑨 𝑩 f → IsHom-op 𝑩 𝑪 g → IsHom-op 𝑨 𝑪 (g ∘ f)
+ ∘-IsHom-op {f}{g} fho gho 𝑓 a = cong g (fho 𝑓 a) ∙ gho 𝑓 (f ∘ a)
+
+ ∘-IsHom : {f : (fst 𝑨) → (fst 𝑩)}{g : (fst 𝑩) → (fst 𝑪)}
+  →         IsHom 𝑨 𝑩 f → IsHom 𝑩 𝑪 g → IsHom 𝑨 𝑪 (g ∘ f)
+ ∘-IsHom {f} {g} fhro ghro = ihr , iho
+  where
+  ihr : IsHom-rel 𝑨 𝑪 (g ∘ f)
+  ihr = ∘-IsHom-rel {f}{g} (fst fhro) (fst ghro)
+
+  iho : IsHom-op 𝑨 𝑪 (g ∘ f)
+  iho = ∘-IsHom-op {f}{g} (snd fhro) (snd ghro)
+
+ ∘-Hom : Hom 𝑨 𝑩  →  Hom 𝑩 𝑪  →  Hom 𝑨 𝑪
+ ∘-Hom (f , fh) (g , gh) = g ∘ f , ∘-IsHom {f}{g} fh gh
+
+
+module _ {α ρ : Level} where
+
+ 𝒾𝒹 : (𝑨 : Structure {α} {ρ} 𝑅 𝐹) → Hom 𝑨 𝑨
+ 𝒾𝒹 _ = id , (λ R a z → z)  , (λ f a → refl)
+
+module _ {α ρᵃ β ρᵇ : Level}
+         (𝑨 : Structure {α}{ρᵃ} 𝑅 𝐹)
+         (𝑩 : Structure {β}{ρᵇ} 𝑅 𝐹) where
+
+ IsMon : ((fst 𝑨) → (fst 𝑩)) → Type (α ⊔ ρᵃ ⊔ β ⊔ ρᵇ)
```

Comment:
I'm really not a big fan of using capitalization as a way to indicate type. I use qualified imports when I need to import things that have the same concepts defined twice. I think this should just be `ismon` (and same for most of the things in this file).

Reply:
Not sure I agree with anti-capitalization... but I do agree about the qualified imports.  And I really like how it's so easy to  rename stuff in cases where you need to import the same concept from different modules.

Reply:
I'm fine with capitalization. I'm not fine with using capitalization as sneaky means to have a slightly different name for "the same thing".

Reply:
Understood.